service/storage_proxy.cc

```diff
@@ -4317,6 +4323,67 @@ storage_proxy::query(schema_ptr s,
     return do_query(s, cmd, std::move(partition_ranges), cl, std::move(query_options));
 }
 
+future<query::forward_result> storage_proxy::dispatch(query::forward_request req)
+{
+    schema_ptr schema = local_schema_registry().get(req.cmd.schema_version);
+    keyspace& ks = _db.local().find_keyspace(schema->ks_name());
+
+    auto generator = query_ranges_to_vnodes_generator(get_token_metadata_ptr(), schema, req.pr);
+    auto ranges = generator(1024);
+    // next_vnode is used to iterate through all vnodes produced by query_ranges_to_vnodes_generator
+    auto next_vnode = [
+        generator = std::move(generator),
+        ranges = std::move(ranges),
+        i = size_t(0)
+    ] () mutable -> std::optional<dht::partition_range> {
+        if (i < ranges.size()) {
+            return {ranges[i++]};
+        } else {
+            ranges = generator(1024);
+            i = 0;
+            if (i >= ranges.size())
+                return {};
+            else
+                return {ranges[i++]};
+        }
+    };
+
+    // group vnodes by assigned endpoint
+    std::map<netw::messaging_service::msg_addr, dht::partition_range_vector> vnodes_per_addr;
+    while (std::optional<dht::partition_range> vnode = next_vnode()) {
+        inet_address_vector_replica_set live_endpoints = get_live_endpoints(ks, end_token(*vnode));
+        auto endpoint_addr = netw::messaging_service::msg_addr{*live_endpoints.begin(), 0};
+        vnodes_per_addr[endpoint_addr].push_back(*vnode);
+    }
+
+    std::optional<query::forward_result> result;
+    return do_with(std::move(vnodes_per_addr), std::move(req), std::move(result),
+        [this] (auto& vnodes_per_addr, auto& req, auto& result) {
+            // forward request to each endpoint and merge results
```

Comment:
Please avoid auto parameters, it makes the thing a template and everything becomes more complicated. In any case you there's no need to use do_with, just make it a coroutine.